# Objective: 

The tweet objects collected from twitter api does not include link to the articles at NYT websites. So this section scrapes those articles from NYT websites. 

In [154]:
import newspaper
from newspaper import Article

https://github.com/huangy22/NewsRecommender/blob/master/4_news_recommendation.ipynb
https://blog.insightdatascience.com/news4u-recommend-stories-based-on-collaborative-reader-behavior-9b049b6724c4

In [155]:
import pandas as pd
import json
def pp_json(json_thing, sort=True, indents=4):
    if type(json_thing) is str:
        print(json.dumps(json.loads(json_thing), sort_keys=sort, indent=indents))
    else:
        print(json.dumps(json_thing, sort_keys=sort, indent=indents))
    return None

In [156]:
ny = pd.read_csv('data/nytimesposts.csv')
df = pd.read_csv('data/retweeterlist.csv')
ny.shape, df.shape

((3226, 7), (3226, 3))

In [157]:
ny.head(2)

Unnamed: 0                      created_at             tweet_id  \
0           0  Mon Feb 18 06:07:07 +0000 2019  1097376949511565312   
1           0  Mon Feb 18 05:49:55 +0000 2019  1097372619848986625   

                                                text                     url  \
0  RT @nytrealestate: After "tidying" your home à...  http://nyti.ms/2FVHq9v   
1  Boys with high levels of inattention later ear...  http://nyti.ms/2FVHq9v   

   retweet_count                                             status  
0             10  {"created_at": "Mon Feb 18 06:07:07 +0000 2019...  
1             39  {"created_at": "Mon Feb 18 05:49:55 +0000 2019...

In [158]:

df.head(2)

Unnamed: 0             tweet_id  \
0           0  1097376949511565312   
1           0  1097372619848986625   

                                      retweeter_list  
0                                                 []  
1  [486097436, 55651793, 191972027, 22681390, 897...

In [159]:
ny.tail(2)

Unnamed: 0                      created_at             tweet_id  \
3224           0  Fri Jan 18 15:14:01 +0000 2019  1086280555468070912   
3225           0  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
3224  RT @NYTStyles: “Sexual health wellness is some...   
3225  Poems often came to Mary Oliver on her walks, ...   

                         url  retweet_count  \
3224  http://nyti.ms/2FVHq9v             77   
3225  http://nyti.ms/2FVHq9v            181   

                                                 status  
3224  {"created_at": "Fri Jan 18 15:14:01 +0000 2019...  
3225  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...

In [160]:
#The rows are not sorted in df due to the skipped rows caused by maxing out the allowed api requests by twitter 
df.tail(2)

Unnamed: 0             tweet_id  \
3224           0  1086572235077439488   
3225           0  1086295933300674562   

                                         retweeter_list  
3224                                                 []  
3225  [2584753063, 2964537632, 1010865833818578944, ...

In [161]:
#merging on 'tweet_id'
df1 = pd.merge(ny, df, how='outer', on=['tweet_id'])


In [162]:
df1.shape, ny.shape

((3226, 9), (3226, 7))

In [163]:
df1.tail(2)

Unnamed: 0_x                      created_at             tweet_id  \
3224             0  Fri Jan 18 15:14:01 +0000 2019  1086280555468070912   
3225             0  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
3224  RT @NYTStyles: “Sexual health wellness is some...   
3225  Poems often came to Mary Oliver on her walks, ...   

                         url  retweet_count  \
3224  http://nyti.ms/2FVHq9v             77   
3225  http://nyti.ms/2FVHq9v            181   

                                                 status  Unnamed: 0_y  \
3224  {"created_at": "Fri Jan 18 15:14:01 +0000 2019...             0   
3225  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...             0   

                                         retweeter_list  
3224                                                 []  
3225  [134479233, 1032816968498339840, 15670312, 101...

In [164]:
#checking that the new merged df is in order
(ny.tweet_id == df1.tweet_id).sum(), ny.shape[0]

(3226, 3226)

In [165]:
#df1.to_csv('data/nytimesposts1.csv')
ny = pd.read_csv('data/nytimesposts1.csv')
ny.shape

(3226, 10)

In [166]:
ny.head(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
0           0             0  Mon Feb 18 06:07:07 +0000 2019   
1           1             0  Mon Feb 18 05:49:55 +0000 2019   

              tweet_id                                               text  \
0  1097376949511565312  RT @nytrealestate: After "tidying" your home à...   
1  1097372619848986625  Boys with high levels of inattention later ear...   

                      url  retweet_count  \
0  http://nyti.ms/2FVHq9v             10   
1  http://nyti.ms/2FVHq9v             39   

                                              status  Unnamed: 0_y  \
0  {"created_at": "Mon Feb 18 06:07:07 +0000 2019...             0   
1  {"created_at": "Mon Feb 18 05:49:55 +0000 2019...             0   

                                      retweeter_list  
0                                                 []  
1  [486097436, 55651793, 191972027, 22681390, 897...

In [167]:
ny.tail(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
3224        3224             0  Fri Jan 18 15:14:01 +0000 2019   
3225        3225             0  Fri Jan 18 15:00:12 +0000 2019   

                 tweet_id                                               text  \
3224  1086280555468070912  RT @NYTStyles: “Sexual health wellness is some...   
3225  1086277077685301249  Poems often came to Mary Oliver on her walks, ...   

                         url  retweet_count  \
3224  http://nyti.ms/2FVHq9v             77   
3225  http://nyti.ms/2FVHq9v            181   

                                                 status  Unnamed: 0_y  \
3224  {"created_at": "Fri Jan 18 15:14:01 +0000 2019...             0   
3225  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...             0   

                                         retweeter_list  
3224                                                 []  
3225  [134479233, 1032816968498339840, 15670312, 101...

In [168]:
#FINDING ROWS THAT DO NOT HAVE RETWEETER INFO
ind_to_drop = []
for i in range(ny.shape[0]): 
    no_of_retweeters = ny['retweeter_list'][i]
    if no_of_retweeters == '[]': 
        ind_to_drop.append(i)
len(ind_to_drop)

615

In [169]:
# DROPPING ALL ROWS FOR TWEETS THAT HAVE NO RETWEETER INFO
ny.drop(ind_to_drop, inplace = True)
ny.reset_index(inplace=True)
ny.drop(['index'], axis=1, inplace = True)
ny.shape

(2611, 10)

In [170]:
#checking if the index has been reset properly or not
ny.tail(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
2609        3223             0  Fri Jan 18 15:30:04 +0000 2019   
2610        3225             0  Fri Jan 18 15:00:12 +0000 2019   

                 tweet_id                                               text  \
2609  1086284595207778305  Speaker Nancy Pelosi postponed an official tri...   
2610  1086277077685301249  Poems often came to Mary Oliver on her walks, ...   

                         url  retweet_count  \
2609  http://nyti.ms/2FVHq9v            214   
2610  http://nyti.ms/2FVHq9v            181   

                                                 status  Unnamed: 0_y  \
2609  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...             0   
2610  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...             0   

                                         retweeter_list  
2609  [1006666433257000962, 1010865833818578944, 595...  
2610  [134479233, 1032816968498339840, 15670312, 101...

In [171]:
#CHECKING FOR NAN VALUES 
ny.isna().any()

Unnamed: 0        False
Unnamed: 0_x      False
created_at        False
tweet_id          False
text              False
url               False
retweet_count     False
status            False
Unnamed: 0_y      False
retweeter_list    False
dtype: bool

In [172]:
# json.loads(ny['status'][6])

In [173]:
#getting url links for each articles from the tweeter post objects and also listing the indexes that did 
#not have proper links 
good_ind_url = []
bad_indexes = []
        
for i in range(ny.shape[0]): 
    print(i)
    try: 
        j = json.loads(ny['status'][i])
        url = j['entities']['urls'][0]['expanded_url']
        good_ind_url.append((i, url))
    except Exception as e:
        bad_indexes.append(i)
        print(e)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [174]:
good_ind_url[0], len(bad_indexes)

((0, 'https://nyti.ms/2EbrJZO'), 2)

In [175]:
ny.shape

(2611, 10)

In [176]:
#updating all the urls  

for i, url in good_ind_url: 
    ny.at[i, 'url'] = url

In [177]:
ny.head(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
0           1             0  Mon Feb 18 05:49:55 +0000 2019   
1           2             0  Mon Feb 18 05:35:18 +0000 2019   

              tweet_id                                               text  \
0  1097372619848986625  Boys with high levels of inattention later ear...   
1  1097368941842837506  What the Opportunity rover saw on Mars https:/...   

                       url  retweet_count  \
0  https://nyti.ms/2EbrJZO             39   
1  https://nyti.ms/2DL2pZ8             93   

                                              status  Unnamed: 0_y  \
0  {"created_at": "Mon Feb 18 05:49:55 +0000 2019...             0   
1  {"created_at": "Mon Feb 18 05:35:18 +0000 2019...             0   

                                      retweeter_list  
0  [486097436, 55651793, 191972027, 22681390, 897...  
1  [794833170, 275942684, 34883086, 7913281211442...

In [178]:
ny.tail(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
2609        3223             0  Fri Jan 18 15:30:04 +0000 2019   
2610        3225             0  Fri Jan 18 15:00:12 +0000 2019   

                 tweet_id                                               text  \
2609  1086284595207778305  Speaker Nancy Pelosi postponed an official tri...   
2610  1086277077685301249  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2609  https://twitter.com/i/web/status/1086284595207...            214   
2610  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  Unnamed: 0_y  \
2609  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...             0   
2610  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...             0   

                                         retweeter_list  
2609  [1006666433257000962, 1010865833818578944, 595...  
2610  [134479233, 1032816968498339840, 15670312, 101...

In [179]:
#DROPPING ALL ROWS FOR TWEETS THAT HAVE NO RETWEETER INFO
ny.drop(bad_indexes, inplace = True)
ny.reset_index(inplace=True)
ny.drop(['index'], axis=1, inplace = True)
ny.shape

(2609, 10)

In [180]:
#checking if the index has been reset properly or not
ny.tail(2)

Unnamed: 0  Unnamed: 0_x                      created_at  \
2607        3223             0  Fri Jan 18 15:30:04 +0000 2019   
2608        3225             0  Fri Jan 18 15:00:12 +0000 2019   

                 tweet_id                                               text  \
2607  1086284595207778305  Speaker Nancy Pelosi postponed an official tri...   
2608  1086277077685301249  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2607  https://twitter.com/i/web/status/1086284595207...            214   
2608  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  Unnamed: 0_y  \
2607  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...             0   
2608  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...             0   

                                         retweeter_list  
2607  [1006666433257000962, 1010865833818578944, 595...  
2608  [134479233, 1032816968498339840, 15670312, 101...

In [181]:
to_drop = ['Unnamed: 0', 'Unnamed: 0_x', 'Unnamed: 0_y']
ny = ny.drop(to_drop, axis=1)

In [182]:
ny.tail(2)

created_at             tweet_id  \
2607  Fri Jan 18 15:30:04 +0000 2019  1086284595207778305   
2608  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
2607  Speaker Nancy Pelosi postponed an official tri...   
2608  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2607  https://twitter.com/i/web/status/1086284595207...            214   
2608  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  \
2607  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...   
2608  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...   

                                         retweeter_list  
2607  [1006666433257000962, 1010865833818578944, 595...  
2608  [134479233, 1032816968498339840, 15670312, 101...

In [11]:
def get_article(url):
    a = Article(url, language='en')
    a.download()
    a.parse()
    return a

In [183]:
import requests
from bs4 import BeautifulSoup
import re

In [184]:
url = 'https://twitter.com/i/web/status/1091029613151748097'
r = requests.get(url).content
soup = BeautifulSoup(r, 'html.parser')

In [185]:

city_details = soup.findAll('div', {'class':"js-tweet-text-container"})
city_details[0].p

<p class="TweetTextSize TweetTextSize--jumbo js-tweet-text tweet-text" data-aria-label-part="0" lang="en">"What does 'Be Mine' even mean if it’s not stamped in red vegetable dye on a lump of sugar, corn syrup, dextrose and glycerin?"<a class="twitter-timeline-link u-hidden" data-expanded-url="https://nyti.ms/2UxzL4S" dir="ltr" href="https://t.co/19ETWjfrAF" rel="nofollow noopener" target="_blank" title="https://nyti.ms/2UxzL4S"><span class="tco-ellipsis"></span><span class="invisible">https://</span><span class="js-display-url">nyti.ms/2UxzL4S</span><span class="invisible"></span><span class="tco-ellipsis"><span class="invisible"> </span></span></a></p>

In [186]:
city_details[0].p.a['href']

'https://t.co/19ETWjfrAF'

In [194]:
#function to update pickle object
import pickle
def pickle_obj(obj, key, path): 
    load_file = open(path, 'rb')
    url_dict = pickle.load(load_file)

    url_dict[key].append(obj)

    update_file = open(path, "wb")
    pickle.dump(url_dict, update_file)
    update_file.close()

In [226]:
import numpy as np
ny['final_url'] = ['0' for i in range(ny.shape[0])]


In [227]:
ny.head(7)

created_at             tweet_id  \
0  Mon Feb 18 05:49:55 +0000 2019  1097372619848986625   
1  Mon Feb 18 05:35:18 +0000 2019  1097368941842837506   
2  Mon Feb 18 05:19:02 +0000 2019  1097364848890007552   
3  Mon Feb 18 04:59:06 +0000 2019  1097359831034486785   
4  Mon Feb 18 04:41:05 +0000 2019  1097355297138581507   
5  Mon Feb 18 04:25:07 +0000 2019  1097351280505339909   
6  Mon Feb 18 04:04:05 +0000 2019  1097345986463256576   

                                                text  \
0  Boys with high levels of inattention later ear...   
1  What the Opportunity rover saw on Mars https:/...   
2  South Asians have a higher death rate from hea...   
3  "True Detective" episode 7 recap: "When Hays d...   
4  Netflix is making a big Oscar push for “Roma,”...   
5  In one of his first interviews after being rel...   
6  "I'm 47, my son is 7 and my daughter is 5, but...   

                                                 url  retweet_count  \
0                            https://nyti.ms/2EbrJZO             39   
1                            https://nyti.ms/2DL2pZ8             93   
2  https://twitter.com/i/web/status/1097364848890...            104   
3  https://twitter.com/i/web/status/1097359831034...             33   
4  https://twitter.com/i/web/status/1097355297138...             62   
5  https://twitter.com/i/web/status/1097351280505...            196   
6  https://twitter.com/i/web/status/1097345986463...            265   

                                              status  \
0  {"created_at": "Mon Feb 18 05:49:55 +0000 2019...   
1  {"created_at": "Mon Feb 18 05:35:18 +0000 2019...   
2  {"created_at": "Mon Feb 18 05:19:02 +0000 2019...   
3  {"created_at": "Mon Feb 18 04:59:06 +0000 2019...   
4  {"created_at": "Mon Feb 18 04:41:05 +0000 2019...   
5  {"created_at": "Mon Feb 18 04:25:07 +0000 2019...   
6  {"created_at": "Mon Feb 18 04:04:05 +0000 2019...   

                                      retweeter_list final_url  
0  [486097436, 55651793, 191972027, 22681390, 897...         0  
1  [794833170, 275942684, 34883086, 7913281211442...         0  
2  [767648943106723840, 275942684, 192261211, 287...         0  
3  [275942684, 21155228, 292394125, 1093559884518...         0  
4  [275942684, 95699657, 894838959536254976, 1500...         0  
5  [1328992454, 16871799, 958839395012014080, 185...         0  
6  [1025419341771104256, 89098742, 120838637, 257...         0

In [228]:
import time



path = "data/urls.pickle"

final_urls = []
good_urls = []
errors = []

url_dict = {'final_urls': final_urls, 'good_urls': good_urls, 'errors': errors}

# f = open(path,"wb")
# pickle.dump(url_dict, f)
# f.close()

count = -1
file_path = 'data/final_urls.csv'
### MAKE EMPTY DF To POPULATE WITH DATA ###########
df1 = pd.DataFrame(columns = ['tweet_id', 'url', 'final_url'])
df1.to_csv(file_path)

for i in range(ny.shape[0]): 

    url = ny['url'][i]
    
    mini_df = pd.DataFrame()
    mini_df['tweet_id'] = [ny['tweet_id'][i]]
    mini_df['url'] = [url]
    
    if 'https://nyti' in url: 
        print(f'url check at index {i}, good url:', url)
        
        mini_df['final_url'] = [url]
        
        ny.at[i, 'final_url'] = url
        
        obj = (i, target_url)
        url_dict['final_urls'].append(obj)
        url_dict['good_urls'].append(i)
        
        time.sleep(1)
    
    else: 
        
        try: 
            r = requests.get(url).content
            soup = BeautifulSoup(r, 'html.parser')
            one = soup.findAll('div', {'class':"js-tweet-text-container"})
            target_url = one[0].p.a['href']

            mini_df['final_url'] = target_url

            ny.at[i, 'final_url'] = target_url

            obj = (i, target_url)
            url_dict['final_urls'].append(obj)
            time.sleep(1)


        except Exception as e:

            print('error at index:', i, 'err_msg:', e)
            print('url check:', url)
 
            url_dict['errors'].append(i)
            mini_df['final_url'] = '0'
            
        
    mini_df.to_csv(file_path, mode='a', header=False)


url check at index 0, good url: https://nyti.ms/2EbrJZO
url check at index 1, good url: https://nyti.ms/2DL2pZ8
url check at index 7, good url: https://nyti.ms/2DN9d8G
url check at index 13, good url: https://nyti.ms/2DJgoip
url check at index 17, good url: https://nyti.ms/2DLWvar
url check at index 19, good url: https://nyti.ms/2DKmFdx
url check at index 20, good url: https://nyti.ms/2EgLtvo
url check at index 37, good url: https://nyti.ms/2SK4eQN
url check at index 43, good url: https://nyti.ms/2SDsZxT
url check at index 68, good url: https://nyti.ms/2EdRp8j
url check at index 71, good url: https://nyti.ms/2DHlECX
url check at index 72, good url: https://nyti.ms/2DGGv9E
url check at index 73, good url: https://nyti.ms/2DIqC2o
url check at index 79, good url: https://nyti.ms/2DJQlrc
url check at index 80, good url: https://nyti.ms/2EcSYDy
url check at index 82, good url: https://nyti.ms/2DIbj9W
url check at index 84, good url: https://nyti.ms/2DSKss3
url check at index 85, good url: h

url check at index 764, good url: https://nyti.ms/2DrFGkK
url check at index 772, good url: https://nyti.ms/2DjCpUG
url check at index 773, good url: https://nyti.ms/2E0cGlD
url check at index 776, good url: https://nyti.ms/2E0CL4d
url check at index 779, good url: https://nyti.ms/2E350z9
url check at index 787, good url: https://nyti.ms/2E1OnEd
url check at index 804, good url: https://nyti.ms/2E0RzzI
url check at index 816, good url: https://nyti.ms/2DZ87rU
url check at index 823, good url: https://nyti.ms/2Dm8Bql
url check at index 828, good url: https://nyti.ms/2E00Rfy
url check at index 831, good url: https://nyti.ms/2DZF9Z6
url check at index 836, good url: https://nyti.ms/2DZozIz
url check at index 844, good url: https://nyti.ms/2DZgCmU
url check at index 851, good url: https://nyti.ms/2DiUSAU
url check at index 855, good url: https://nyti.ms/2E0DvGB
url check at index 862, good url: https://nyti.ms/2Dn266D
url check at index 878, good url: https://nyti.ms/2DYcHGS
url check at i

url check at index 1388, good url: https://nyti.ms/2DOuOyI
url check at index 1396, good url: https://nyti.ms/2Ux8HCw
url check at index 1403, good url: https://nyti.ms/2UAagzL
url check at index 1412, good url: https://nyti.ms/2DOa36v
url check at index 1424, good url: https://nyti.ms/2UAEOkZ
url check at index 1425, good url: https://nyti.ms/2UweGrh
url check at index 1427, good url: https://nyti.ms/2DPqHTn
url check at index 1430, good url: https://nyti.ms/2DOGtOc
url check at index 1432, good url: https://nyti.ms/2DNlafY
url check at index 1438, good url: https://nyti.ms/2Uwrmys
url check at index 1443, good url: https://nyti.ms/2UA7GK9
url check at index 1451, good url: https://nyti.ms/2Uy9vqR
url check at index 1465, good url: https://nyti.ms/2DKXtoo
url check at index 1469, good url: https://nyti.ms/2DMuJMg
url check at index 1475, good url: https://nyti.ms/2UqPxhM
url check at index 1489, good url: https://nyti.ms/2DKFqPt
url check at index 1504, good url: https://nyti.ms/2DN7X

url check at index 2225, good url: https://nyti.ms/2FQdATk
url check at index 2227, good url: https://nyti.ms/2FWB3Sy
url check at index 2233, good url: https://nyti.ms/2FX4Afh
url check at index 2246, good url: https://nyti.ms/2FVzFQt
url check at index 2248, good url: https://nyti.ms/2FSg3MW
url check at index 2250, good url: https://nyti.ms/2DtWOHV
url check at index 2251, good url: https://nyti.ms/2FRNRcX
url check at index 2255, good url: https://nyti.ms/2DtvoC9
url check at index 2258, good url: https://nyti.ms/2FRSFPO
url check at index 2260, good url: https://nyti.ms/2FPj00M
url check at index 2266, good url: https://nyti.ms/2FUAAAF
url check at index 2274, good url: https://nyti.ms/2DoXVsg
url check at index 2278, good url: https://nyti.ms/2FVNDBV
url check at index 2285, good url: https://nyti.ms/2Dq9P5p
url check at index 2290, good url: https://nyti.ms/2DqSoBx
url check at index 2291, good url: https://nyti.ms/2DpO1H4
url check at index 2299, good url: https://nyti.ms/2Dri9

In [231]:
ny.shape

(2609, 8)

In [230]:
# ny.to_csv('data/nytimesposts2.csv')

In [293]:
ny2 = pd.read_csv('data/nytimesposts2.csv')
ny2.tail(2)

Unnamed: 0                      created_at             tweet_id  \
2607        2607  Fri Jan 18 15:30:04 +0000 2019  1086284595207778305   
2608        2608  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
2607  Speaker Nancy Pelosi postponed an official tri...   
2608  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2607  https://twitter.com/i/web/status/1086284595207...            214   
2608  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  \
2607  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...   
2608  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...   

                                         retweeter_list  \
2607  [1006666433257000962, 1010865833818578944, 595...   
2608  [134479233, 1032816968498339840, 15670312, 101...   

                    final_url  
2607  https://t.co/HdTA1Kgcoa  
2608  https://t.co/rxb84GYSPg

In [294]:
final_urls = pd.read_csv('data/final_urls.csv')
final_urls.shape

(2609, 4)

In [242]:
final_urls.tail(2)

Unnamed: 0             tweet_id  \
2607           0  1086284595207778305   
2608           0  1086277077685301249   

                                                    url  \
2607  https://twitter.com/i/web/status/1086284595207...   
2608  https://twitter.com/i/web/status/1086277077685...   

                    final_url  
2607  https://t.co/HdTA1Kgcoa  
2608  https://t.co/rxb84GYSPg

In [295]:
print(errors)

[638, 1019, 1022, 1023, 1024, 1025, 1026, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1040]


In [296]:
#checking the error urls
for i in errors:
    print(ny2.iloc[i, 4])

https://twitter.com/i/web/status/1094732765508132865
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2DUZsXs
http://nyti.ms/2TFVaZe


In [298]:
# ny2.drop(errors, inplace = True)
# ny2.reset_index(inplace=True)
# ny2.tail(2)

level_0  index  Unnamed: 0                      created_at  \
2590     2607   2607        2607  Fri Jan 18 15:30:04 +0000 2019   
2591     2608   2608        2608  Fri Jan 18 15:00:12 +0000 2019   

                 tweet_id                                               text  \
2590  1086284595207778305  Speaker Nancy Pelosi postponed an official tri...   
2591  1086277077685301249  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2590  https://twitter.com/i/web/status/1086284595207...            214   
2591  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  \
2590  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...   
2591  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...   

                                         retweeter_list  \
2590  [1006666433257000962, 1010865833818578944, 595...   
2591  [134479233, 1032816968498339840, 15670312, 101...   

                    final_url  
2590  https://t.co/HdTA1Kgcoa  
2591  https://t.co/rxb84GYSPg

In [299]:
# ny2.drop(['index', 'Unnamed: 0' ], axis=1, inplace = True)
# ny2.tail(2)

level_0                      created_at             tweet_id  \
2590     2607  Fri Jan 18 15:30:04 +0000 2019  1086284595207778305   
2591     2608  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
2590  Speaker Nancy Pelosi postponed an official tri...   
2591  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2590  https://twitter.com/i/web/status/1086284595207...            214   
2591  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  \
2590  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...   
2591  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...   

                                         retweeter_list  \
2590  [1006666433257000962, 1010865833818578944, 595...   
2591  [134479233, 1032816968498339840, 15670312, 101...   

                    final_url  
2590  https://t.co/HdTA1Kgcoa  
2591  https://t.co/rxb84GYSPg

PART II: UPDATING LINKS TO CSV FILE

## Validating links with the text from nylatest df

In [300]:
# Validating links with the text
n = 10
print('\n', ny2['final_url'][n])
print('\n', ny2['url'][n])
print('\n', ny2['text'][n])

n = 20
print('\n', ny2['final_url'][n])
print('\n', ny2['url'][n])
print('\n', ny2['text'][n])

n = 1556

print('\n', ny2['final_url'][n])
print('\n', ny2['url'][n])
print('\n', ny2['text'][n])


 /hashtag/MeTooUyghur?src=hash

 https://twitter.com/i/web/status/1097322571643736064

 A #MeTooUyghur social media campaign aims to pressure China on missing Uighurs: "Show me that my father is alive an… https://t.co/bPjLC1j5Er

 https://nyti.ms/2EgLtvo

 https://nyti.ms/2EgLtvo

 Raised in the United States? Our 2013 quiz shows how y’all, youse and you guys talk. https://t.co/Fsxddg5qGC

 https://t.co/GkN0TfmRF1

 https://twitter.com/i/web/status/1090694904668450816

 The judge presiding over more than 1,500 federal opioid lawsuits had hoped to settle them by now. But the behemoth… https://t.co/UOhGj01PDi


Some of the final_url are still faulty 

In [301]:
#finding all the faulty urls and setting value to any
final_urls = url_dict['final_urls']
bad_url = []
for i in range(ny2.shape[0]):
    url = ny2['final_url'][i]
    if not 'https' in url: 
        print(i, url)
        ny2.at[i, 'final_url'] = np.NaN
        bad_url.append(i)

10 /hashtag/MeTooUyghur?src=hash
41 /david_marchese
66 /katierogers
67 /peterbakernyt
69 /parul_sehgal
100 /peterbakernyt
115 /katierogers
167 /sebmodak
184 /david_marchese
186 /david_marchese
187 /david_marchese
200 /npfandos
212 /katierogers
218 /dealbook
243 /brycecovert
266 /PatriciaMazzei
267 /ScottCacciola
275 /kevinroose
296 /ScottCacciola
314 /dealbook
336 /NickKristof
401 /dealbook
420 /zinoman
469 /poniewozik
493 /dealbook
515 /bxchen
535 /parul_sehgal
549 /hashtag/MeToo?src=hash
554 /MaggieAstor
562 /kzernike
563 /kzernike
564 /kzernike
565 /kzernike
575 /kzernike
590 /hashtag/savehakeem?src=hash
607 /hashtag/Grammys?src=hash
610 /joncaramanica
628 /nytimesmusic
645 /DrJenGunter
649 /UpshotNYT
709 /Wesley_Morris
819 /dealbook
865 /rcallimachi
866 /rcallimachi
867 /rcallimachi
868 /rcallimachi
869 /rcallimachi
871 /tripgabriel
876 /mattfleg
884 /mattfleg
888 /GiniaNYT
902 /dealbook
925 /fmanjoo
964 /MarkLandler
967 /VVFriedman
970 /dougmillsnyt
1002 /hashtag/YearofthePig?src=

In [305]:
c = 0
for i in bad_url: 
    c +=1
    print(c, ny2['url'][i])

1 https://twitter.com/i/web/status/1097322571643736064
2 https://twitter.com/i/web/status/1097197767523557377
3 https://twitter.com/i/web/status/1097095838583717889
4 https://twitter.com/i/web/status/1097088294444298241
5 https://twitter.com/i/web/status/1097077013591011328
6 https://twitter.com/i/web/status/1096907109764161536
7 https://twitter.com/i/web/status/1096846704689713153
8 https://twitter.com/i/web/status/1096582459804844032
9 https://twitter.com/i/web/status/1096508892266332170
10 https://twitter.com/i/web/status/1096505971298500608
11 https://twitter.com/i/web/status/1096504176828387329
12 https://twitter.com/i/web/status/1096457885712875523
13 https://twitter.com/i/web/status/1096417108001476609
14 https://twitter.com/i/web/status/1096393728963551232
15 https://twitter.com/i/web/status/1096299331043364864
16 https://twitter.com/i/web/status/1096189847604285440
17 https://twitter.com/i/web/status/1096186334635532288
18 https://twitter.com/i/web/status/1096152119621705728
1

In [306]:
to_drop = ny2[ny2.final_url.isna()]

In [307]:
#DROPPING ALL THE BAD URLS
ny2.drop(to_drop.index, axis = 0, inplace =True )
ny2.reset_index(inplace = True)
print(ny2.shape)
ny2.tail(2)

(2467, 10)


index  level_0                      created_at             tweet_id  \
2465   2590     2607  Fri Jan 18 15:30:04 +0000 2019  1086284595207778305   
2466   2591     2608  Fri Jan 18 15:00:12 +0000 2019  1086277077685301249   

                                                   text  \
2465  Speaker Nancy Pelosi postponed an official tri...   
2466  Poems often came to Mary Oliver on her walks, ...   

                                                    url  retweet_count  \
2465  https://twitter.com/i/web/status/1086284595207...            214   
2466  https://twitter.com/i/web/status/1086277077685...            181   

                                                 status  \
2465  {"created_at": "Fri Jan 18 15:30:04 +0000 2019...   
2466  {"created_at": "Fri Jan 18 15:00:12 +0000 2019...   

                                         retweeter_list  \
2465  [1006666433257000962, 1010865833818578944, 595...   
2466  [134479233, 1032816968498339840, 15670312, 101...   

                    final_url  
2465  https://t.co/HdTA1Kgcoa  
2466  https://t.co/rxb84GYSPg

In [291]:
# ny2.to_csv('data/nytposts_readyforarticle.csv')

SUMMARY: 
1. Final number of rows after dropping bad urls: 2467
2. After getting artilce check for text = https://www.nytimes.com/tips (tips)
3. Next up collect article text- get_articles.ipynb